File: SwiftUI/Guides/Models/Planet.swift
import Foundation

struct Planet: Decodable {
    let _id: String
    let hasRings: Bool
    let isArchived: Bool
    let mainAtmosphere: [String]
    let name: String
    let orderFromSun: Int
    let planetId: String
    let surfaceTemperatureC: Temperature
    
    struct Temperature: Decodable {
        let max: Double?
        let mean: Double
        let min: Double?
    }
}

extension Planet {
    init(value: [String: Any?]) {
        _id = value["_id"] as! String
        hasRings = value["hasRings"] as! Bool
        isArchived = value["isArchived"] as! Bool
        mainAtmosphere = value["mainAtmosphere"] as! [String]
        name = value["name"] as! String
        orderFromSun = value["orderFromSun"] as! Int
        planetId = value["planetId"] as! String
        
        // Safely unwrap the temperature dictionary
        if let temp = value["surfaceTemperatureC"] as? [String: Any] {
            // Now we can safely access the temperature values
            let max = temp["max"] as? Double
            let min = temp["min"] as? Double
            let mean = temp["mean"] as? Double ?? 0.0  // Provide default value since mean is non-optional
            surfaceTemperatureC = Temperature(
                max: max,
                mean: mean,
                min: min
            )
        } else {
            // Provide default values if temperature data is missing
            surfaceTemperatureC = Temperature(
                max: nil,
                mean: 0.0,
                min: nil
            )
        }
    }
}

File: Android/app/src/main/java/com/ditto/guides/models/Planet.kt
package com.ditto.guides.models

import androidx.annotation.Keep
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Keep
@Serializable
data class Planet(
    @SerialName("_id") val id: String,
    val hasRings: Boolean,
    val isArchived: Boolean,
    val mainAtmosphere: List<String>,
    val name: String,
    val orderFromSun: Int,
    val planetId: String,
    val surfaceTemperatureC: Temperature) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Planet) return false

        return id == other.id &&
                hasRings == other.hasRings &&
                isArchived == other.isArchived &&
                mainAtmosphere == other.mainAtmosphere &&
                name == other.name &&
                orderFromSun == other.orderFromSun &&
                planetId == other.planetId &&
                surfaceTemperatureC == other.surfaceTemperatureC
    }

    override fun hashCode(): Int {
        var result = id.hashCode()

        result += 31 * result + hasRings.hashCode()
        result += 31 * result + isArchived.hashCode()
        result += 31 * result + mainAtmosphere.hashCode()
        result += 31 * result + name.hashCode()
        result += 31 * result + orderFromSun.hashCode()
        result += 31 * result + planetId.hashCode()
        result += 31 * result + surfaceTemperatureC.hashCode()

        return result
    }

    companion object {
        fun fromMap(value: Map<String, Any?>): Planet {
            return Planet(
                id = value["_id"] as String,
                hasRings = value["hasRings"] as Boolean,
                isArchived = value["isArchived"] as Boolean,
                mainAtmosphere = (value["mainAtmosphere"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList(),
                name = value["name"] as String,
                orderFromSun = value["orderFromSun"] as Int,
                planetId = value["planetId"] as String,
                surfaceTemperatureC = (value["surfaceTemperatureC"] as? Map<*, *>)?.let { temp ->
                    Temperature(
                        max = (temp["max"] as? Number)?.toDouble(),
                        mean = (temp["mean"] as? Number)?.toDouble() ?: 0.0,
                        min = (temp["min"] as? Number)?.toDouble()
                    )
                } ?: Temperature(mean = 0.0)
            )
        }
    }

}

@Keep
@Serializable
data class Temperature(
    val max: Double? = null,
    val mean: Double,
    val min: Double? = null) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Temperature) return false

        return max == other.max &&
                mean == other.mean &&
                min == other.min
    }

    override fun hashCode(): Int {
        var result = mean.hashCode()
        max?.let {
            result += 31 * result + it.hashCode()
        }
        min?.let {
            result += 31 * result + it.hashCode()
        }
        return result
    }
}